KIALI-2618 Saves graph filterState values

josejulio opened the pull request:
** Describe the change **

Saves in local storage the filterState values from redux:

 edgeLabelMode, findValue, graphType, hideValue, showCircuitBreakers,
 showFindHelp, showLegend, showMissingSidecars, showNodeLabels,
 showSecurity, showServiceNodes, showTrafficAnimation,
 showUnusedNodes, showVirtualServices

** Issue reference **

https://issues.jboss.org/browse/KIALI-2618

** Backwards compatible? **

[ ] Is your pull-request introducing changes in behaviour? 
Is persiting filterState values to localstorage


** Screenshot **

No UI changes

josejulio commented on src/reducers/GraphDataState.ts:
```diff
@@ -39,127 +40,171 @@ export const INITIAL_GRAPH_STATE: GraphState = {
+      return state;
   }
-
-  return newState;
```
to clarify, this is needed, but regardless this change, we should include it.
This reducer was returning a new state even when there wasn't any change or what is worst, even on unrelated actions.
This was confusing / making fail the "persist: hydrate" action.

josejulio commented on src/reducers/GraphDataState.ts:
```diff
@@ -7,6 +7,7 @@ import { EdgeLabelMode } from '../types/GraphFilter';
 import { GraphType } from '../types/Graph';
 import { GraphFilterActions } from '../actions/GraphFilterActions';
 import { DagreGraph } from '../components/CytoscapeGraph/graphs/DagreGraph';
+import { updateState } from '../utils/Reducer';
```
The old way of this reducer was wrong [1] (emphasis mine):

> We return the previous state in the default case. **It's important to return the previous state for any unknown action**

Furthermore, redux uses the returned state to know if it has changed (I recall reading this somewhere, and quickly googled a blogpost [2] that shows it [3])

By always returning a new object, we are signaling a state change for **every** action, and when I say every, I mean **every**, not only the graphData reducers.


[1] https://redux.js.org/basics/reducers
[2] https://www.freecodecamp.org/news/why-redux-needs-reducers-to-be-pure-functions-d438c58ae468/
[3] https://cdn-media-1.freecodecamp.org/images/lxQkKAq1x7nWzokdsojlcF9gjhQi3ZUNM38Z

israel-hdez replied:
I was meaning the old way of forming the new object. That updateState function is only a wrapper for the spread operator...

josejulio replied:
Ah, you made me write such a long text for nothing :smile:

israel-hdez replied:
While writing that reply, I imagined you like Homer in this image: https://amp.businessinsider.com/images/54296c3decad04d80d31ea15-480-219.jpg
...and me being Bart


Not true :P

josejulio commented:
> Code looks good. Switching from/to the graph looks good, settings are maintained. The "show on graph" link also maintains settings. Editing the settings that are also bookmarkable in the URL seems to work, the URL setting takes precedence. F5 maintains the settings with one glitch, the graph find/hide values don't take effect until the first refresh. Please take a look and see if that can be resolved.
> 

Fixed


> One unrelated issue, I see that F5 is resetting the duration. I think this is a regression. If it's convenient to fix as part of this PR then good, otherwise let's create a new Jira.

I found the problem and fixed it

jshaughn commented:
> Graph Find/Hide dialog, then a browser refresh closes the dialog.

I think it's OK, probably better, if the dialog closes.  As long as the toggling continues to work as expected, it's fine by me.

josejulio commented:
> I think it's OK, probably better, if the dialog closes. As long as the toggling continues to work as expected, it's fine by me.

going to merge now. We can change this behavior in a separate PR if needed.
